### PR TITLE
Rotate with right-drag in qap and if focused module not using it

### LINKED ISSUE #10757: click+drag rotation correction as a global shortcut
There have been some (predictable) comments complaining about the loss of the crop & rotate module from darktable 3.8.

Personally the only thing that I really miss is the ability to right-click+drag to set the rotation from within the crop module. Going back and forth from crop to rotate & perspective and back again isn't the best workflow.

Would it be possible to activate this functionality using a keyboard shortcut (accessible from outside of the module) and have this change the parameter in the rotate & perspective module?

### LINKED ISSUE #12222: RFC - rotation in quick access panel
When the rotation slider is added into the quick access panel we get the slider and we can adjust the rotation but this does not activate the way to adjust using the right-click to draw the line on the canvas directly.

Would be nice to have this supported when the module is activated from the quick access panel.

### PR DESCRIPTION
**Pull request by dterrahe**
fixes #12222
fixes #10757

This always let's the rotate & perspective module process right-click & drags, unless something else already claims it. So it works in QAP and when most other modules are focused, but _not_ when crop is active (because then right-click resets the crop), or when color-picking or when editing masks (even when not above a mask, for safety).

Also require that a right-drag needs to be over a certain distance before an angle will be calculated (and shown) and rotation applied when released. So a simple right click (without mouse move) will be ignored.

Adds a toast with the amount of the adjustment and the new angle (for when rotation widget or panel are hidden).

### CODE REVIEW

**Comment by TurboGit** [comment]
> Since apparently this is the single most important issue that stops people from even considering dt and makes all other progress irrelevant (https://librearts.org/2023/01/year-in-preview/).

Indeed, how can I have been able to used dt all those years!!!

**Comment by dterrahe** [comment]
> Indeed, how can I have been able to used dt all those years!!

I bet you are downloading LR right now!

**Comment by elstoc** [comment]
I'm think I'm ok with the approach described in this PR (I haven't tested it) but might prefer it be activated with a key- or button-press.

I would guess, from various half-remembered discussions in the past, that what most users might like/expect is a button in the interface (maybe on the bottom bar) that is easily-discoverable and allows click+drag to set rotation (could be with left-click if it's activated via a button), with maybe something similar for crop. Then that button could be assigned to a shortcut for even-quicker-ness. I expect there are people who expect a right-click context menu and _might_ accidentally activate rotation when they try to activate such a thing (probably they'd only make that mistake once though).

**Comment by TurboGit** [comment]
> > Indeed, how can I have been able to used dt all those years!!
> 
> I bet you are downloading LR right now!

Yes, and you'll have to find another maintainer for dt :) Don't want to use such bad software :)

**Comment by dterrahe** [comment]
> Yes, and you'll have to find another maintainer for dt :)

It's all fun and games, but now you're scaring the kids! Don't even say that in jest.

**Comment by dterrahe** [comment]
> I would guess [...] that what most users might like/expect

> I expect there are people who expect a right-click context menu

I am particularly bad at imagining what the typical user (the one undaunted by 15 sliders in the _diffuse or sharpen_ module but easily flummoxed by a right-click) might expect so I'm going to stay out of the way of the last item on the todo list (as indicated by the Request For Adoption). I find little joy these days in implementing functionality that I'd rather not use myself, even if the powers that be demand it. Instead I'm going to carry this little patch forward in my private collection (together with other abominations, like #12705 and #12624) until a better solution comes along. I'm even thinking of publishing a mini-fork of my own with just those few additions for the unsatisfied with a deathwish. Working title: _Ancil_(lary).

**Comment by TurboGit** [comment]
Call me fool but I like this RFA and find it useful. I won't fight for it but I think it would be nice and will fit perfectly in dt workflow. If right-click is problematic maybe we could add a key (r + right-click?).

**Comment by dterrahe** [comment]
Round about now might be the right time for those elusive UX people to show up and spend some quality time here.

I don't know any, especially that won't either run away screaming or patiently tell us that we have to throw _everything_ away and start from scratch so that we can build something our grandmother can use. The existing impression to the contrary, the project actually doesn't have _any_ development expertise or capacity that would allow them to tackle a problem of this scale. If we scrape everything together we maybe have 1 1/2 people that really understand gtk. And spending a humongous amount of time on something only to then be told you've now made things _too_ configurable and you should really block people from doing crazy things in their own personal setup is kinda demotivating.

The kind of input that _would_ be useful would be to suggest relatively small improvements that mostly fit in the currently implemented approach, so they won't require changing (and breaking) too much code or habit. Or to step in to prevent things from getting worse. To be able to provide _that_, they have to spend time getting to know the current state a little (warts and all). Less quick and easy and likely no shiny end product to point at with pride. "Fewer warts" just doesn't sound as sexy and the reward will mostly be people pointing at the _remaining_ warts.

In short, knowing the _problems_ (786 issues and counting) doesn't necessarily mean we know the best _answers_.

**Comment by TurboGit** [comment]
In full agreement with @dterrahe here. The UI certainly need some polishing. But we need a team for that with a long term view and commitment. There is nothing that can be done in a month when we talk about a full redesign of the UI. We need a clear goal and a step by step approach to reach the goal. If you read that and you are a UI guy, please speak we us and let's try to find a motivated team 2 or 3 people min.

**Comment by dterrahe** [comment]
I'm not sure if we are saying the same thing. With the current team and budget I believe "a full redesign of the UI" will never work. Not only do 2 or 3 people need multiple years to do it, everybody else has to put up with everything being (temporarily) broken at the end. If that's the plan, then starting from a totally different base (vkdt...) might lead to a more satisfactory end result. But I'm sure there are feasible fixes that can be made that won't fix the fundamental nature of dt, or its learning curve, but at least could make it less frustrating to use for those who persist in learning all the technical skills.

**Comment by TurboGit** [comment]
@dterrahe : Maybe you are correct and I'm underestimating the amount of work or maybe I'm just seeing something less ambitious. Not sure, but I'm anyway pretty sure that a full redesign will never happen :)

**Comment by dterrahe** [comment]
> maybe I'm just seeing something less ambitious

Having a manageable list of paper cuts would be a good place to begin. So not a mockup where everything has moved and all functionality redefined, but small annoyances that really hinder productivity _even_ for the experienced (or that cause people to walk away in disgust before they get there). Anything that looks like it would need fundamental changes or that would require too much other functionality to _also_ change to avoid confusing inconsistencie should not be on this list. To start with, this could just be a triage of the whole issue list, which has now become too unwieldy to invite involvement. Somebody with fresh eyes (who hasn't skimmed that list many times already) could really help here.

**Comment by AlicVB** [comment]
For me a full redesign is neither realist, nor good. I would opt for small steps, and contrary to what you say, I think that we already have some people around that keeps proposing small enhancements, sometimes good, sometimes less pertinent...

Back to the subject...
- iirc some times ago we have proposed to have a button for the ashift module (well c&r in fact) that would display a line on screen with both end moveable (like in gratuated density). That would solve the discoverability... and allow the use in QAP (well that will some change there). That feature would "duplicate" the actual right-click one which is more direct for experience users.
- tbh I'm not a big fan of a "global" right click, as imho we'll have too many "collision" with other features... not to speak that new "on canvas" features for modules... will make a regression by "removing" this one !
- if we go that path anyway, we'll certainly have to completely disable it if some drawing masks are on : imho it's complex enough to click on the right position to act on the right mask ;)

**Comment by prokoudine** [comment]
@dterrahe To me, cropping and rotation activation are at the top of discoverability issues list in darktable.

For cropping, darktable (like a few more photography applications) draws from the GIMP model where you have side/corners areas for dragging. GIMP's model however isn't perfect. It's very convenient in terms of low effort when tweaking the cropping frame, because you have a large area for grabbing. But it has this discoverability issue where you have to figure out that you need to draw the cropping frame first. And to do that, you need to hover the image in the viewport. Otherwise there are no visual hints as nothing changed in the viewport (for darktable, it literally says "working" when you enable the module, then nothing in the image changes). I watched GIMP users fail at this time and time again.

What darktable could do is, immediately upon enabling the crop module, render the entire cropping frame the size of the image, with all four corner areas for grabbing visible and side areas displayed on hover. That is both visual feedback to an action and a hint what to do.

Doing that would not present a consistency issue. When you enable the graduated density module, it immediately displays the on-canvas widget for controlling the effect. Darktable does not ask the user to add that control. Similarly, when you enable vignetting, you do not have to add concentric circles, they are already there as soon as you enable the module.

More or less the same applies to rotation.

That's one side of the story. The other side of the story is that cropping and rotation are essential tools that need to be easily available. However, in the default setup they are in the middle of a long list of modules in the "Technical" group by default.

I completely understand that this type of software tends to blow up in number of features over years. So the "easy" solution is to add shortcuts, create quick access panels, and search boxes. Whereupon you will eventually have to ask yourself: are cropping/rotation the kind of a tool that need to be located with the help of a search box? Or are they so essential that they deserve to be visible and easily accessible? Like, being in the toolbar?

Again, placing them in the quick access panel might sound like an easy solution that doesn't require adding buttons to the top or the bottom toolbar. But then you have to deal with two facts:

1. Enabling the rotate/perspective module in the quick access panel does not enable the "draw a line in the viewport" feature, you have to go to the full module first.

2. For some inexplicable reason (worth a separate report) darktable defaults to the Grading group of modules when switching to the darkroom view, as if WB, exposure, composition, and other basic properties were assumed to be correct by default, and so it's now time to add bling. This makes rotation effectively out of sight.

Normally, the top toolbar in the darkroom view would be the place where you could put these activation buttons. However, right now, the top toolbar in the darkroom view is occupied by the filter machinery. I'm not sure why.

darktable's UI imposes a particular workflow: import, sort/cull, fix basic properties like WB and tonality, fix composition/lens stuff, color-grade, apply effects. So when you go to the darkroom view, it should be generally assumed that the contents of the filmstrip are already the right kind of photos that are ready for processing. So the top toolbar is currently used for things that appear to be in conflict with assumed workflow, while essential tools are hidden in the middle of a long list of options.

**Comment by dterrahe** [comment]
@prokoudine thanks for the suggestions. This is exactly the kind of feedback I'd like to see more of. It would be nice to have a more complete list in one go (so that the solution to one of these papercuts isn't going to preclude solutions to others or to identify that a series of small but not negligible changes is actually one larger one that solves many problems at once, but better). The details/justification could then maybe be hidden in a collapsible section; quite often for the real papercuts these are obvious anyway. So we probably don't require a whole report about the default group. Maybe the github _projects_ or _discussions_ functionality could be used for this.

The highlighted discoverability problem with crop should hopefully be very easily solvable. See if #13377 gets closer and please suggest improvements. Fully agreed that the model used by dt for cropping is probably not ideal, but a complete overhaul (especially after the recent spilt-off of rotate) will probably not find any takers.

The quick access panel was definitely _not_ intended as a _solution_ (or even improvement) specifically for crop and/or rotate. It was meant to fulfill the demand for a do-it-all single basic processing module by making it configurable and therefore not just limited to beginners. But the clash it created with the per-module-overlay approach that dt is based on hasn't really been addressed. One step forward, a shuffle back, because there's simply not the luxury to put a team on a multi-year project to release the next all-singing-dancing AND tested new version.

Yes, the top bar (filter and config) doesn't make as much sense in the darkroom as it does in the lighttable (and there is a good chance that many users simply hide it in the darkroom, especially if they don't use the filmstrip either). But changing it likely requires changes in many places to provide the needed functionality _somewhere_ and since the current users have their preferences, these discussions often go nowhere. Everybody's proposal is the obviously right one _to them_. So then whoever takes on the big job knows not to expect any thanks (happy customers are quiet ones, unhappy customers not so much).

As you can see in this PR, even something as simple as rotation isn't so simple when you add 10 opinions. But I would really like you to stick around and be involved in further stages of refining actionable proposals (and also _testing_; it would be great if we found some people who are serious about testing. It is well known that developers are the worst testers of their own work since they develop blind spots almost faster than features).

**Comment by dterrahe** [comment]
Two refinements:
- when editing masks, don't respond to right-drag with rotation, because there is a risk a user doing this would accidentally delete shapes. This can be overridden with "force pan & zoom with mouse".
- When right-dragging, require a minimum distance before calculation of the angle is accurate enough to perform a rotation. This prevents a user who might think right-click would pop up a menu from accidentally rotating. Also add a toast specifying by how much the rotation is adjusted and what the new rotation is (in case the module is not expanded or even the whole panel is hidden). This would also help a confused user. Also fixes a small artifact where the drawn line end nodes get ridiculously large when zoomed.

I think this now reasonable safe and useful. Maybe there are other states that (sometimes) don't do anything with right-click but still would want to prevent it from being handled by rotate; these can be dealt with in a similar way as masks (i.e. only a small change would be needed). So the only other reason I can see why we would not want to implement this is that we'd want right-drag to be used for something _else_. I don't see that happening immediately, but if ever, then presumably this could be easily removed again along with changes required to rotate&persp and graduatednd, among others.

So I'll submit this as non-draft, reformulate the OP and see what happens.

**Comment by dterrahe** [comment]
@turbogit I have fixed all the buggy interactions with other modules that I noticed. right-drag rotate now works in QAP and while any module is focused that doesn't use it and while not editing masks or color picking. In those situations, for example in the crop module, you can still force it (along with pan and zoom) by holding A (the shortcut action has been renamed).

Please let me know if you spot any issues or if you have particular reservations.

Side node; while working on this, I was somewhat irked by the repetitive and inefficient use of `dt_dev_get_zoom_scale` and `dt_dev_get_pointer_zoom_pos` and the associated calls and calculations, but that will have to be a separate cleanup at some point.

**Comment by TurboGit** [comment]
@dterrahe : Thanks, I'll test and report.

**Review by TurboGit — APPROVED** [review]
Works for me and a nice usability enhancement. I'll merge for more field testing. Thanks!

**Comment by wpferguson** [comment]
@dterrahe thanks!  This saves me lots of clicks.
